cpp/src/arrow/array/array_base.h

```diff
@@ -232,6 +232,14 @@ class ARROW_EXPORT Array {
   /// \return DeviceAllocationType
   DeviceAllocationType device_type() const { return data_->device_type(); }
 
+  /// \brief Return the statistics of this Array
+  ///
+  /// This just delegates to calling statistics on the underlying ArrayData
+  /// object which backs this Array.
+  ///
+  /// \return const ArrayStatistics&
+  const ArrayStatistics& statistics() const { return data_->statistics; }
```

Comment:
Should statistics be stored in memory together with every `ArrayData` instance?

Reply:
Another problem with this is that statistics are derived data and `ArraData` is mutable when manipulated directly, so any mutation of `ArrayData` will have to consider the consequences to the derived statistics.

Lazily-computed `null_count_` is a source of bugs and complexity for this reason. IMO statistics should be (1) computed or (2) carried from a file readers (like Parquet's) as something on the side.

Reply:
> Should statistics be stored in memory together with every `ArrayData` instance?

If it's not desired, we can avoid it by using `std::shared_ptr<ArrayStatistics>` or something.

> Another problem with this is that statistics are derived data and `ArraData` is mutable when manipulated directly, so any mutation of `ArrayData` will have to consider the consequences to the derived statistics.
>
> Lazily-computed `null_count_` is a source of bugs and complexity for this reason. IMO statistics should be (1) computed or (2) carried from a file readers (like Parquet's) as something on the side.

How about attaching the statistics read by a file reader to `Array` (not `ArrayData`) directly?

Reply:
> How about attaching the statistics read by a file reader to Array (not ArrayData) directly?

Makes more sense. Perhaps even attach it only to the typed array classes and not `Array` itself.

Reply:
I've added `Array::SetStatistics()` for this.
I wanted to pass statistics by constructor to prevent changing statistics after construction but I didn't do it. Because our constructors already require many arguments. For example, `PrimiriveArray` has 6 arguments:

https://github.com/apache/arrow/blob/797ca30c2bc634f3352d44b1b9bf0bc017a3e426/cpp/src/arrow/array/array_base.h#L272-L275

`SetStatistics()` isn't thread safe because `std::shared_ptr::operator=()` isn't thread safe. So users must set statistics before parallel processing.

What do you think about this API?

Reply:
There shouldn't be a public statistics setter in `Array` — an immutable class: `SetStatistics` should be `protected` just like `SetData`.

Leave the existing constructor as is, extend the `protected` constructor that takes `ArrayData`:

```diff
  protected:
   PrimitiveArray() : raw_values_(NULLPTR) {}

   void SetData(const std::shared_ptr<ArrayData>& data) {
     this->Array::SetData(data);
     raw_values_ = data->GetValuesSafe<uint8_t>(1, /*offset=*/0);
   }

-  explicit PrimitiveArray(const std::shared_ptr<ArrayData>& data) { SetData(data); }
+  PrimitiveArray(const std::shared_ptr<ArrayData>& data,
+                 std::shared_ptr<ArrayStats> stats) {
+    SetData(data);
+    SetStats(std::move(stats));
+ }

   const uint8_t* raw_values_;
```

Then you change `BooleanArray` constructor that takes `ArrayData` to also take statistics:

```diff
-  explicit BooleanArray(const std::shared_ptr<ArrayData>& data);
+  BooleanArray(const std::shared_ptr<ArrayData>& data,
+               std::shared_ptr<ArrayStats> stats);
```

Reply:
The stats `shared_ptr` should become an additional (but optional) parameter to `MakeArray`

```cpp
std::shared_ptr<Array> MakeArray(const std::shared_ptr<ArrayData>& data) {
  std::shared_ptr<Array> out;
  ArrayDataWrapper wrapper_visitor(data, &out);
  DCHECK_OK(VisitTypeInline(*data->type, &wrapper_visitor));
  DCHECK(out);
  return out;
}
```

Reply:
OK. I'll choose the changing only `ArrayData` constructors approach.

Reply:
Done:

* I've added `Array::Array(const std::shared_ptr<ArrayData>& data, const std::shared_ptr<ArrayStatistics>& statistics)` that calls `SetData()` and `SetStatistics()` and use it instead of direct `SetData()` call (as much as possible) in sub arrays.
* I've added `ValidateData()` and it's called by `SetData()` to use `Array::Array(data, statistics)` in sub arrays.
* I've unified `const std::shared_ptr<ArrayData>& data` argument and `std::shared_ptr<ArrayData> data` argument in sub array's constructors to `const &` because related codes use `const &`.

There are many small diffs for the change. Review may be a bit difficult.

Reply:
Changes that affect the overall design of `Array` should probably extracted to a separate PR so they can be reviewed more carefully. It seems very risky to decouple `ValidateData` from `SetData`.

Reply:
It makes sense.
I've opened https://github.com/apache/arrow/pull/43273 that only include `arrow::ArrayStatistics`.

We can keep using this PR for discussing `arrow::ArrayStatitics` related APIs because this PR still includes not only `arrow::ArrayStatistics` but also Apache Parquet integration example.